pkg/cri/config/config.go

```diff
@@ -71,12 +71,24 @@ type Runtime struct {
 	// be loaded from the cni config directory by go-cni. Set the value to 0 to
 	// load all config files (no arbitrary limit). The legacy default value is 1.
 	NetworkPluginMaxConfNum int `toml:"cni_max_conf_num" json:"cniMaxConfNum"`
+
+	// Snapshotter setting snapshotter at runtime level instead of making it as a global configuration.
+	// An example use case is to use devmapper or other snapshotters in Kata containers for performance and security
+	// while using default snapshotters for operational simplicity.
+	// See https://github.com/containerd/containerd/issues/6657 for details.
+	Snapshotter string `toml:"snapshotter" json:"snapshotter"`
 }
 
 // ContainerdConfig contains toml config related to containerd
 type ContainerdConfig struct {
 	// Snapshotter is the snapshotter used by containerd.
 	Snapshotter string `toml:"snapshotter" json:"snapshotter"`
+
+	// PullImageRuntimeAnnotationKey specifies the annotation key to get runtime type from PullImageRequest.SandboxConfig.Annotations
+	// Ideally we would want the runtime to be passed in as a field in PullImageRequest, which requires kubelet support.
+	// See https://github.com/containerd/containerd/issues/6657
+	PullImageRuntimeAnnotationKey string `toml:"pull_image_runtime_annotation_key" json:"pullImageRuntimeAnnotationKey"`
```

Comment:
Not sure we want this. Maybe discussed in a separate PR.

Reply:
Where is this annotation getting set?

Reply:
@dmcgowan annotations will be set in pod sandbox config, see the following config as an example:

```
{
  "metadata": {
    "name": "busybox-sandbox-devmapper",
    "namespace": "default",
    "attempt": 1,
    "uid": "hdishd83djaidwnduwk28bcsb"
  },
  "log_directory": "/tmp",
  "linux": {
  },
  "annotations": {
    "myruntimekey": "my-runtime"
  }
}
```

Reply:
> Not sure we want this. Maybe discussed in a separate PR.

@AkihiroSuda I agree this annotation is not as good as having the runtime passed in from the PullImageRequest. The problem we are seeing is that PullImage does not have a way of passing the runtime info to CRI, so CRI will unpack images with default snapshotter. When containers are created using different snapshotter, the images will be unpacked again, this essentially uncompressed layers twice and added significant latency.

I think we had a discussion here https://github.com/containerd/containerd/issues/6657#issuecomment-1086383548 , but I'm definitely open to better alternatives.

This annotation seems to be a relatively no-harm option to address such issue without modifying CRI API and update kubelet. In the future if kubelet can pass the runtime during pull time, we can deprecate it.

Reply:
Adding a note, kubelet seems already has an annotation key `kubernetes.io/runtimehandler` https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/images/image_manager_test.go#L266 to pass the runtime handler during PullImage. It was also mentioned in @kevpar's comment https://github.com/containerd/containerd/issues/6657#issuecomment-1066478023. Unfortunately, Kubelet does not set this annotation, instead, user need to add the annotations by themselves.

So we can either use this hardcoded key `kubernetes.io/runtimehandler` or make the key configurable. Since `kubernetes.io/runtimehandler` is not yet in any protocol/spec/agreement yet, I feel making it a config offers more flexibility. 

@AkihiroSuda @kevpar what do you think?

Reply:
suggest just make the key a const string

Reply:
pls mark the const as needing to be removed.. upon a KEP adding the runtime handler to the image services

Reply:
Const SGTM.

The annotation should be probably prefixed with "io.kubernetes.cri." ?

https://github.com/containerd/containerd/blob/main/pkg/cri/annotations/annotations.go

Reply:
Agree with just hardcoding the annotation key for now.

For the annotation format, it's a little unclear what is expected. [K8s specifies](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set) that annotations should be in the `foo.com/anno-name` form. Whereas [OCI specifies](https://github.com/opencontainers/runtime-spec/blob/main/config.md#annotations) that they should be in `com.foo.anno-name` form. Ideally, it seems like CRI plugin should convert between these two forms. However, in practice both seem to be used today.

Reply:
Throwing in another idea seems to be more transparent to users and also introduces less annotation and future backward compatibility break concerns. We can leverage the existing `criServer.sandboxNameIndex` to look up sandboxID from `PullImageRequst.PodSandboxConfig` and therefore to get the sandbox runtime.

![image](https://user-images.githubusercontent.com/28563006/167701205-96fe1caa-f7ca-44c5-8f5b-a6eb452bbaf6.png)

This approach seems to be a better approach to me, @kevpar @mikebrow @AkihiroSuda does it sound good to you?

Reply:
That approach would require the pod to always be started before the image is pulled. I don't know if that is a safe assumption.

Reply:
We will check the PodSandboxConfig.Metadata to be not nil and name not empty. If there's no sandbox associated with the PullImage (e.g. crictl pull), then using the default snapshotter seems to be a proper behavior. 

To me it's the same situation with annotations, because in order to get annotation, the pod should already exist (not necessarily created).

The only exception is when pull image with a PodSandboxConfig that has metadata set but not stored in containerd, which I feel it's ok to fallback to the default snapshotter.

I'm fine with either approach, let's see if others will chime in for more points, if not I can update to the const annotation key approach seems most of the folks are onboard with a const annotation key.

Reply:
> That approach would require the pod to always be started before the image is pulled. I don't know if that is a safe assumption.

It's definitely not a safe assumption. Preloading and importing images won't work in this fashion.  Agree with the compromise of handler (annotation for now) being nil means use default though and in cases where and admin doesn't want the default behavior they can leave the default out.

Reply:
Agreed and I've updated accordingly.